src/firmware.cpp

```diff
@@ -192,109 +160,100 @@ static CommandError ultrasound_read(int commandId, String argument) {
+  serialWrite(requestID, '>', String(distanceInt));
 
   return OK;
 }
 
-static CommandError get_version(int commandId, String argument) {
-  serialWrite(commandId, '>', FIRMWARE_VERSION);
+static CommandResponse version(int requestID, String argument) {
+  serialWrite(requestID, '>', FIRMWARE_VERSION);
   return OK;
 }
 
-static const CommandHandler commands[] = {
-  CommandHandler("help", &run_help, "show information"),
-  CommandHandler("led", &led, "control the debug LED (on/off)"),
-  CommandHandler("servo", &servo, "control a servo <num> <width>"),
-  CommandHandler("version", &get_version, "get firmware version"),
-  CommandHandler("gpio-write", &write_pin, "set output from GPIO pin"),
-  CommandHandler("gpio-read", &read_pin, "get digital input from GPIO pin"),
-  CommandHandler("analogue-read", &analogue_read, "get all analogue inputs"),
-  CommandHandler("ultrasound-read", &ultrasound_read, "read an ultrasound sensor <trigger-pin> <echo-pin>"),
-};
+static CommandResponse writePin(int requestID, String argument) {
+  String pinIDArg = pop_option(argument);
+  String pinStateArg = pop_option(argument);
 
-static void serialWrite(int commandId, char lineType, const String& str) {
-    if (commandId != 0) {
-        Serial.write('@');
-        Serial.print(commandId, DEC);
-        Serial.write(' ');
-    }
+  if (argument.length() || !pinIDArg.length() || !pinStateArg.length()) {
+    return COMMAND_ERROR("Bad Arguments");
+  }
 
-    Serial.write(lineType);
-    Serial.write(' ');
+  int pin = pinIDArg.toInt();
 
-    Serial.println(str);
+  if (pin < 2 || pin > 13) {
+    return COMMAND_ERROR("pin must be between 2 and 13");
+  }
+
+  switch(pinStateArg.charAt(0)){
+    case 'H':
+      pinMode(pin, OUTPUT);
+      digitalWrite(pin, HIGH);
+      break;
+    case 'L':
+      pinMode(pin, OUTPUT);
+      digitalWrite(pin, LOW);
+      break;
+    case 'P':
+      pinMode(pin, INPUT_PULLUP);
+      break;
+    case 'Z':
+      pinMode(pin, INPUT); // High Impedance
+      break;
+    default:
+      return COMMAND_ERROR("Unknown pin mode");
+    
+  }
+  return OK;
 }
 
-static void dispatch_command(int commandId, const class CommandHandler& handler, const String& argument) {
-  auto err = handler.run(commandId, argument);
+// Process the commands and execute them.
+
+static const CommandHandler commands[] = {
+  CommandHandler('A', &analogueRead), // Read the analogue pins
+  CommandHandler('L', &led), // Control the debug LED (H/L)
+  CommandHandler('R', &readPin), // Read a digital pin <number>
+  CommandHandler('S', &servo), // Control a servo <num> <width>
+  CommandHandler('U', & ultrasoundRead), // Read an ultrasound distance
+  CommandHandler('V', &version), // Get firmware version
+  CommandHandler('W', &writePin), // Write to or  a GPIO pin <number> <state>
+};
+
+static void dispatch_command(int requestID, const class CommandHandler& handler, const String& argument) {
+  auto err = handler.run(requestID, argument);
```

Comment:
as it's been renamed to `CommandResponse`, might as well update this to `response`

Reply:
Fixed in a79124d